### PR TITLE
ci: publish workflow

### PR DESCRIPTION
**Pull request by Shinigami92**
This is a GitHub action workflow that let's maintainers of Faker run a npm publish process, so that the package `@faker-js/faker` will be published to npm registry.

It is respecting the version and dist-tag by reading the package.json.

The process will be:
1. Someone create a normal release PR
2. The release PR gets merged into next branch
3. Someone hits the publish workflow in the github actions menu `// <- this was done locally before`
4. After the release succeeded, someone will normally create the GitHub Release

It requires that one of the maintainers create a new access token on NPM for the `@faker-js/faker` package, or for the `@faker/*` orga-scope.
This then needs to be configured in GitHub as a secret with the name `NPM_AUTH_TOKEN`, so it can be picked up by the CI.

Note that `publish` will run the `prepublishOnly` script that is configured in the `package.json`

This workflow will NOT "craft" a release, nor will it create tags/releases on github.
It just and only automates the publishing to npm.

Hint: the workflow is tested for 3 months now in https://github.com/salsita/node-pg-migrate

### CODE REVIEW

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      - name: Prepare
+        run: pnpm install --frozen-lockfile
+
+      - name: Set publishing config
+        run: pnpm config set '//registry.npmjs.org/:_authToken' "${NODE_AUTH_TOKEN}"
+        env:
+          NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
+
+      - name: Publish
+        run: pnpm publish --access public --tag latest
```
Comment (Shinigami92) [inline .github/workflows/publish.yml]: Rewrite this to respect alpha/beta dist-tag

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+          node-version: 20
+          cache: 'pnpm'
+
+      - name: Prepare
+        run: pnpm install --frozen-lockfile
+
+      - name: Set publishing config
+        run: pnpm config set '//registry.npmjs.org/:_authToken' "${NODE_AUTH_TOKEN}"
+        env:
+          NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
```
Comment (ST-DDT) [inline .github/workflows/publish.yml]: IMO we should keep the variable name consistent.

```suggestion
          NPM_AUTH_TOKEN: ${{ secrets.NPM_AUTH_TOKEN}}
```

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -0,0 +1,36 @@
+name: Publish
+
+on:
+  workflow_dispatch:
+
+jobs:
```
Comment (ST-DDT) [inline .github/workflows/publish.yml:9]: Please configure the jobs permission, same as in other workflows.

File: .github/workflows/publish.yml (hunk 2 above)
Comment (Shinigami92) [reply .github/workflows/publish.yml]: Please note that this is coming from the official GitHub documentation and I don't know if there is anything specific behind that

https://docs.github.com/en/actions/publishing-packages/publishing-nodejs-packages#publishing-packages-to-the-npm-registry

Comment (ST-DDT) [reply .github/workflows/publish.yml]: > I don't know if there is anything specific behind that

Infinite copy and paste + npm uses the env variable directly, but pnpm does not.
The secret name is entirely up to the user.

**Review by ST-DDT — DISMISSED** [review]

File: .github/workflows/publish.yml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+      - name: Set publishing config
+        run: pnpm config set '//registry.npmjs.org/:_authToken' "${NPM_AUTH_TOKEN}"
+        env:
+          NPM_AUTH_TOKEN: ${{ secrets.NPM_AUTH_TOKEN }}
+
+      - name: Publish
+        run: |
+          PACKAGE_DIST_TAG=$(node -e "console.log(/^\d+\.\d+\.\d+(\-(\w+)\.\d+)$/.exec(require('./package.json').version)?.[2] || 'latest')")
+          pnpm publish --access public --tag $PACKAGE_DIST_TAG
```
Comment (ST-DDT) [inline .github/workflows/publish.yml:39]: Maybe add another tag if the branch is `next`.

Release-Checklist:

> - [ ] Set `next` npm dist-tag

Comment (xDivisionByZerox) [reply .github/workflows/publish.yml:39]: Good point. [The only tags we currently have, are](https://www.npmjs.com/package/@faker-js/faker?activeTab=versions):

- latest (the latest current major version)
- stable (the current major version that has been out for at least one week without any incidents)
- old (the latest previous major version)
- next (the latest next major version in prerelease)

File: .github/workflows/publish.yml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+      - name: Set next dist-tag
+        run: |
+          PACKAGE_VERSION=$(node -e "console.log(require('./package.json').version)")
+          pnpm dist-tag add @faker-js/faker@$PACKAGE_VERSION next
```
Comment (ST-DDT) [inline .github/workflows/publish.yml:44]: IMO we should tag it with next by default and set the latest only when it is a stable release:

```suggestion
          PACKAGE_VERSION=$(node -e "console.log(require('./package.json').version)")
          pnpm publish --access public --tag next
		  if [[ "$PACKAGE_VERSION" != *"-"* ]]; then
            pnpm dist-tag add @faker-js/faker@$PACKAGE_VERSION latest
          fi
```

Not tested.

File: .github/workflows/publish.yml [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+      - name: Prepare
+        run: pnpm install --frozen-lockfile
+
+      - name: Set publishing config
+        run: pnpm config set '//registry.npmjs.org/:_authToken' "${NPM_AUTH_TOKEN}"
+        env:
+          NPM_AUTH_TOKEN: ${{ secrets.NPM_AUTH_TOKEN }}
+
+      - name: Publish
```
Comment (ST-DDT) [inline .github/workflows/publish.yml:36]: According to our release checklist, we should remove the "next branch warning" before the publish.

File: .github/workflows/publish.yml (hunk 5 above)
Comment (Shinigami92) [reply .github/workflows/publish.yml:44]: No, we are extracting the dist-tag out of the version field and then e.g. alpha, beta or whatever is used
This is default tagging behavior guided to semver

File: .github/workflows/publish.yml (hunk 6 above)
Comment (Shinigami92) [reply .github/workflows/publish.yml:36]: I'm not sure what this has to do with this PR 🤔

Comment (ST-DDT) [reply .github/workflows/publish.yml:36]: confused: IMO This PR is intended to partially automate the release process, so it should actually contain the same steps as the manual workflow?!
If we don't, we have to permanently remove that section from the readme.

File: .github/workflows/publish.yml (hunk 5 above)
Comment (ST-DDT) [reply .github/workflows/publish.yml:44]: confused: But we don't have alpha or beta tags!? Or did we remove them after stable release?

Comment (Shinigami92) [reply .github/workflows/publish.yml:44]: Oh good that we found out that we misused that before 🫤 
- https://www.npmjs.com/package/vite?activeTab=versions
- https://www.npmjs.com/package/typescript?activeTab=versions
- https://www.npmjs.com/package/react?activeTab=versions

just some examples that use e.g. beta dist-tags (which are not my own packages)

File: .github/workflows/publish.yml (hunk 6 above)
Comment (Shinigami92) [reply .github/workflows/publish.yml:36]: wait, we ever did remove that banner for a release?! I was not aware of that
if I e.g. look in https://www.npmjs.com/package/@faker-js/faker/v/8.4.1 or https://www.npmjs.com/package/@faker-js/faker/v/9.0.0-alpha.0 we did not do that
Only in the newest release and I was not really part of that

IMO we should just remove that banner completely (in a separate PR using `docs: *`)
But if you want to you can also leave this open until thursday

Comment (ST-DDT) [reply .github/workflows/publish.yml:36]: FFR: The comment was added for this issue:

- https://github.com/faker-js/faker/issues/1869

Comment (Shinigami92) [reply .github/workflows/publish.yml:36]: After a not so good sleep last night, I decided (at least for my self) that I'm not interested in manipulating repository code from a workflow and so also dont like to work further on that topic especially not from this PR, as the scope and intend was something different from my side.
I just wanted to provide a workflow that can be used to release the "current" state of next branch without having to use credentials manually.

So feel free to either provide a follow-up PR (my preference) or jump into this PR and take it over (not my preference).

**Comment by ST-DDT** [comment]
Superseded by https://github.com/faker-js/faker/pull/2991 as requested [here](https://github.com/faker-js/faker/pull/2591#discussion_r1652048402).

- https://github.com/faker-js/faker/pull/2991

**Pull request closed without merging**